### ISSUE TITLE
No unitless values

### ISSUE THREAD

**Issue by wazaraki** [stylelint/stylelint#3208]
> Describe the issue. Is it a bug or a feature request (new rule, new optio

I tried to configure `property-unit-whitelist` & `property-unit-blacklist` to flag invalid unitless values. It didn't work

> Which rule, if any, is this issue related to?

`property-unit-whitelist`, `property-unit-blacklist`,

> What CSS is needed to reproduce this issue?

```css
.a {
  margin-bottom: 32; /* <- should error */
}
```

> What stylelint configuration is needed to reproduce this issue?

Any of these won't flag errors in the code snippet above:

```json
{
  "rules": {
    "declaration-property-unit-blacklist": {
            "margin-bottom": []
    },
    "declaration-property-unit-whitelist": {
            "margin-bottom": ["px", "%", "etc"]
    }
  }
}
```

> Which version of stylelint are you using?

`9.1.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint 'src/**/*.css'`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

Unitless values on *width, *height, margin*, border*, etc, to be flagged as errors."

> What actually happened (e.g. what warnings or errors you are getting)?

`declaration-property-unit-whitelist` & `declaration-property-unit-blacklist` don't seem to work to flag undesired unitless values. I couldn't find another way to accomplish this.

**Comment by hudochenkov** [comment]
Thanks for the report and for using the template.

There are no rules covered this case in stylelint. But there is [stylelint-csstree-validator](https://github.com/csstree/stylelint-validator) plugin for checking CSS validity according to specs.

**Comment by kristerkari** [comment]
Could this be reopened @hudochenkov ?

stylelint-csstree-validator only supports CSS (no preprocessors), and there should be are rule to handle this specfic use case without validating other syntax related things.

**Comment by hudochenkov** [comment]
I'm afraid it's against our [criteria for inclusion](https://stylelint.io/developer-guide/rules/#criteria-for-inclusion). Or it's within our criteria, but it'll need to rebuilt `stylelint-csstree-validator` with extra to handle non-standard syntaxes. The best bet here will be to create a plugin if someone feels like it.

**Comment by kristerkari** [comment]
>  I'm afraid it's against our criteria for inclusion. 

What would make it not apply to the criteria?

> it'll need to rebuilt stylelint-csstree-validator with extra to handle non-standard syntaxes.

But `stylelint-csstree-validator` is validating a whole syntax. This issue was to prevent skipping the units for properties that don't support them.

Also, bulding a validator that validates syntax for CSS, Sass, Less, Stylus etc. is really complicated and I don't see that as a good option.

**Comment by hudochenkov** [comment]
Sorry, I confused.

> This issue was to prevent skipping the units for properties that don't support them.

This is exactly what `stylelint-csstree-validator` does if I understand correctly. Also, [it knows](https://github.com/csstree/stylelint-validator/blob/ae68168bcd3e498f549715e4443ea41563b20a3c/syntax-extension/index.js#L10-L14) about non-standard syntax which could be in property value.

**Comment by kristerkari** [comment]
> This is exactly what `stylelint-csstree-validator` does if I understand correctly. Also, [it knows](https://github.com/csstree/stylelint-validator/blob/ae68168bcd3e498f549715e4443ea41563b20a3c/syntax-extension/index.js#L10-L14) about non-standard syntax which could be in property value.

I don't think that you got my point. `stylelint-csstree-validator` validates *the whole syntax*, and because it does not properly support preprocessors, there are a lot of bugs related to the validation of the syntax when Sass or Less is used.

If there would be a rule to validate the unitless values, then it would much simpler to make it work correctly as it only has to care about the property value. It could only support CSS and skip values with interpolation or other non-standard syntax. That rule would be useful for both CSS and preprocessors, without trying to validate syntax that is out of its scope

**Comment by jeddy3** [comment]
> then it would much simpler to make it work correctly as it only has to care about the property value

@kristerkari It would be fantastic if stylelint could validate these unitless values out-of-the-box, but I'm not sure it's possible to do so robustly without incorporating a lexer that validates against the [property-value pair syntaxes defined by W3C](https://github.com/mdn/data/blob/master/css/properties.json).

We use PostCSS because it supports very loose syntax. This has helped us support non-standard syntaxes. As you know with your work on `stylelint-scss` :)

We're kind of stuck in the middle between loose syntax and spec validation, which is why we tried to be pragmatic when jotting down our [VISION](https://github.com/stylelint/stylelint/blob/master/VISION.md#possible-errors):

> Invalid code is best handled by emerging dedicated tools e.g. csstree - a language parser with syntax validation. As a stop-gap, while these tools mature, provide invalid code rules for the simplest of cases.

I'm not sure validating property-value pairs to check for unitless values is a simple case.

---

> The best bet here will be to create a plugin if someone feels like it.

I think @hudochenkov is right. One way forward is to create a stop-gap plugin. A plugin that will allow experimentation by not being tied to stylelint releases. We can then see what it takes to validate property-value pairs.

However, it could just start with a few common properties like `margin` and `padding` and check those. It'll need to account for things like CSS Variables (e.g. `var(--spacing)`). And, as you said, it'd ignore properties or values that contain interpolation or other non-standard syntax.

We can always revisit its inclusion in stylelint itself at a later date.

**Comment by kristerkari** [comment]
Thanks @jeddy3 

I agree that creating a plugin might be the best option as the rule has some complexity involved. I'm just not agreeing on trying to solve the issue of validating units by using a validator that validates the whole syntax. 

The way that I see that the rule could be implemented as simple as possible:

1) Get a list of CSS properties that accept `<length>` values. e.g. from here https://github.com/mdn/data/blob/master/css/properties.json
2) Add support for checking if the value has unit or not, but ignore CSS vars,  functions, interpolation etc.
3) Add support for checking CSS vars and functions after having the basic support working.

**Comment by kristerkari** [comment]
Actually I have a better idea:

1) Get a list of CSS properties that accept <length> values. e.g. from here https://github.com/mdn/data/blob/master/css/properties.json
2) Use csstree to validate only the value of the declaration: https://github.com/csstree/csstree/blob/master/docs/parsing.md#context

**Comment by jeddy3** [comment]
> Use csstree to validate the property and value for those props without validating anything else: 

I'm pretty sure that's how the csstree stylelint plugin [does it](https://github.com/csstree/stylelint-validator/blob/ae68168bcd3e498f549715e4443ea41563b20a3c/index.js#L42-L45).

Limiting the plugin to properties that accept length values will simplify things. I think it's a good idea.
